### PR TITLE
use async callbacks for eldoc

### LINKED ISSUE #113: frequent Sync request timed out documentHighlights
I keep getting this error message so frequently. What could be the cause?

Here is a copy of the backtrace that appears when it happens: 
```
Debugger entered--Lisp error: (error "Sync request timed out documentHighlights")
  signal(error ("Sync request timed out documentHighlights"))
  error("Sync request timed out %s" "documentHighlights")
  tide-send-command-sync("documentHighlights" (:file "~/projects/app/src/App.ts" :line 1 :offset 1 :filesToSearch ("~/projects/app/src/App.ts")))
  tide-command:documentHighlights()
  tide--hl-identifier()
  tide-hl-identifier()
  tide--hl-identifiers-function()
  apply(tide--hl-identifiers-function nil)
  timer-event-handler([t 0 0 100000 t tide--hl-identifiers-function nil idle 0])
```

### PR DESCRIPTION
**Pull request by ananthakumaran**
This removes another sync call and will probably fix #113 and could help with #133
\cc @josteink

### CODE REVIEW

File: tide.el [hunk 1]
```diff
@@ -358,8 +360,7 @@ LINE is one based, OFFSET is one based and column is zero based"
          (payload (concat encoded-command "\n")))
     (process-send-string (tide-current-server) payload)
     (when callback
-      (puthash request-id (cons (current-buffer) callback) tide-response-callbacks)
-      (accept-process-output nil 0.01))))
+      (puthash request-id (cons (current-buffer) callback) tide-response-callbacks))))
```
Comment (josteink) [inline tide.el]: Is there any particular reason for this statement to be removed? Was it redundant and not needed? Does it get in the way for async code-patterns?

**Comment by josteink** [comment]
Sorry for the slow response. I didn't catch the CC and assumed someone else with more time would gloss over this :)

A few questions (apart from the one in the diff): From what I can tell, your basic change here is converting lots of synchronous functions and code-paths into asynchronous ones. There's no other functional changes, right?

For this to impact #113, I'm guessing that async code-paths do not suffer the same time-out issues as synchronous ones does? Is that correct? If so, this looks like a reasonable fix to me.

As for assisting #133, I assume the new helper-function `tide-eldoc-maybe-show` is meant to give us a new mechanism to incrementally cycle through signatures? Is that correct, or are there other changes you've done w.r.t. that particular issue?

Otherwise it all looks good to me, and from what I can tell eldoc still works like it used to, so I don't see anything broken :)

File: tide.el (hunk 1 above)
Comment (ananthakumaran) [inline tide.el]: This adds bit of uncertainity. with accept-process-output, the callback could get executed sync(if the server responds within 0.01) or asyc. This was not a problem in other use cases. For eldoc, the `tide-eldoc-function` always returns nil. If the callback gets executed synchronously, then the value returned by tide-eldoc-function will overwrite the value set by callback. Removing `accept-process-output` will make sure the callback always get executed after the `tide-eldoc-function` returns.

**Comment by ananthakumaran** [comment]
> For this to impact #113, I'm guessing that async code-paths do not suffer the same time-out issues as synchronous ones does? Is that correct? If so, this looks like a reasonable fix to me.

That's my guess also

> As for assisting #133, I assume the new helper-function tide-eldoc-maybe-show is meant to give us a new mechanism to incrementally cycle through signatures? Is that correct, or are there other changes you've done w.r.t. that particular issue?

If we decide to cycle through the signatures, this might be useful, because we have inverted the control and can make changes to the content shown by the eldoc.

**Comment by josteink** [comment]
> That's my guess also

If so, and everything else works, I guess there's no harm in trying and seeing if it does indeed work as expected.

Should we ask @johnbendi to verify before committing these changes? Or do we consider the sync-to-async a general improvement which may improve responsiveness, so we want them regardless?

>  we have inverted the control and can make changes to the content shown by the eldoc

I missed that bit, but that definitely sounds useful. I'm all for it.

Either way, all in all, LGTM. Merge at will.

**Pull request merged**